connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectorRestartApiIntegrationTest.java

```diff
@@ -0,0 +1,425 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.integration;
+
+import org.apache.kafka.connect.storage.StringConverter;
+import org.apache.kafka.connect.util.clusters.EmbeddedConnectCluster;
+import org.apache.kafka.test.IntegrationTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.TestRule;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static org.apache.kafka.connect.integration.MonitorableSourceConnector.TOPIC_CONFIG;
+import static org.apache.kafka.connect.runtime.ConnectorConfig.CONNECTOR_CLASS_CONFIG;
+import static org.apache.kafka.connect.runtime.ConnectorConfig.KEY_CONVERTER_CLASS_CONFIG;
+import static org.apache.kafka.connect.runtime.ConnectorConfig.TASKS_MAX_CONFIG;
+import static org.apache.kafka.connect.runtime.ConnectorConfig.VALUE_CONVERTER_CLASS_CONFIG;
+import static org.apache.kafka.connect.runtime.TopicCreationConfig.DEFAULT_TOPIC_CREATION_PREFIX;
+import static org.apache.kafka.connect.runtime.TopicCreationConfig.PARTITIONS_CONFIG;
+import static org.apache.kafka.connect.runtime.TopicCreationConfig.REPLICATION_FACTOR_CONFIG;
+import static org.apache.kafka.connect.runtime.WorkerConfig.CONNECTOR_CLIENT_POLICY_CLASS_CONFIG;
+import static org.apache.kafka.connect.runtime.WorkerConfig.OFFSET_COMMIT_INTERVAL_MS_CONFIG;
+import static org.apache.kafka.connect.util.clusters.EmbeddedConnectClusterAssertions.CONNECTOR_SETUP_DURATION_MS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+
+/**
+ * Test connectors restart API use cases.
+ */
+@Category(IntegrationTest.class)
+public class ConnectorRestartApiIntegrationTest {
```

Comment:
How long does it take for all of the tests in this IT to run, say locally? IIUC, we're setting up a new Kafka cluster and Connect cluster for every test. How difficult would it be to share the same Kafka and Connect cluster across most of the tests?

Reply:
its on average 7-11sec per test (the ones with 35 sec are due to waiting in NOOP cases)
![image](https://user-images.githubusercontent.com/29556518/122268900-42763900-cea2-11eb-8647-e417265a234d.png)

Let me see if I can find a way to share and if it reduces the time.

Reply:
Yeah, that's already a good portion of the total time to run the Connect unit and integration tests. It'd be great if we could avoid increasing the build times by that much.

Reply:
@rhauch  this was a good find, I just refactored the code, and this saved 2 min

![image](https://user-images.githubusercontent.com/29556518/122320107-52fcd280-cee7-11eb-9e14-3c6f9af3a917.png)

Reply:
Excellent! IIUC, the tests are only using 2 differently-sized clusters (1 node and 4 nodes), which is good because it minimizes the startup time. Can you confirm that statement is true? It seems like there are three total methods that take more than 30 seconds -- any reduction in duration on any of those three will have a very direct impact.

Reply:
I went with a different sleep duration and that cut the times by 1minute 15 sec 
![image](https://user-images.githubusercontent.com/29556518/122449965-132bfe80-cf6c-11eb-8658-e2a389163455.png)

https://github.com/apache/kafka/pull/10822/commits/b60fb53a139dbd44c247ac1a4e2c927ef42ddc99.
I chose 5sec because locally most the test that sets up the cluster takes around 5 sec to do it and  most test after they setup the cluster finishes in <300ms so request propagation and restart isnt taking long but thats my local machine so I am doing 10 times the wait in case cloud or other developer machines are slow.  @rhauch Could you please review to see if you agree with the fix as another idea is to reduce it to 1 sec?

Reply:
The new times after checking that API response didn't return any RESTARTING state is 13 sec and 10 sec out of it is in bringing up kafka cluster.

![image](https://user-images.githubusercontent.com/29556518/122629722-ac4e3880-d084-11eb-8742-8cdaf703ba99.png)